### PR TITLE
CFE-3237 Copy modules from masterfiles to workdir on hub during update

### PR DESCRIPTION
**Pull request by Lex-2008**
Since now they are managed by masterfiles.

#

### CODE REVIEW

File: cfe_internal/update/update_policy.cf [hunk 1]
```diff
@@ -236,6 +236,16 @@ bundle agent cfe_internal_update_policy_cpv
       action => u_immediate,
       classes => u_if_repaired("validated_updates_ready");
 
+    am_policy_hub::
+
+      "$(sys.workdir)$(const.dirsep)modules"
```
Comment (nickanderson) [inline cfe_internal/update/update_policy.cf:241]: I don't think that only the hub should do this. If we are moving to having MPF put modules in place, I think that should happen on each host.

File: cfe_internal/update/update_policy.cf [hunk 2]
```diff
@@ -236,16 +236,18 @@ bundle agent cfe_internal_update_policy_cpv
...
 
-    !am_policy_hub::
+    any::
 
       "$(sys.workdir)$(const.dirsep)modules"
-      comment => "Always update modules files on client side",
+      comment => "Update modules from masterfiles",
       handle => "cfe_internal_update_policy_files_update_modules",
-      copy_from => u_rcp("$(modules_dir_source)", @(update_def.policy_servers)),
+      copy_from => u_cp("$(master_location)$(const.dirsep)modules"),
```
Comment (vpodzime) [inline cfe_internal/update/update_policy.cf:244]: It should still copy the package modules from the policy server(s), right?

File: cfe_internal/update/update_policy.cf [hunk 3]
```diff
@@ -236,16 +236,18 @@ bundle agent cfe_internal_update_policy_cpv
...
       depth_search => u_recurse("inf"),
       perms => u_m("755"),
```
Comment (nickanderson) [inline cfe_internal/update/update_policy.cf:246]: Will this be the right perms on windows?

File: cfe_internal/update/update_policy.cf (hunk 2 above)
Comment (nickanderson) [reply cfe_internal/update/update_policy.cf:244]: Probably it should first copy modules from the policy server(s), then it should place the modules that are vendored by the mpf.

Comment (vpodzime) [reply cfe_internal/update/update_policy.cf:244]: ? I thought the package modules will be part of MPF and thus under `/var/cfengine/masterfiles/` and thus updated like everything else.

Comment (Lex-2008) [reply cfe_internal/update/update_policy.cf:244]: Right. Somewhere else policy copies from `/var/cfengine/masterfiles/` on hub to `/var/cfengine/inputs` on agent. Here we should be copying from `/var/cfengine/inputs/modules` (not masterfiles!) to `/var/cfengine/modules` - and do it on all machines.

@nickanderson : is it okay to do it in two steps or is it better to copy to `/var/cfengine/modules` from `/var/cfengine/masterfiles/modules` at the same time as copying to `/var/cfengine/inputs` from `/var/cfengine/masterfiles` on hub?

Comment (nickanderson) [reply cfe_internal/update/update_policy.cf:244]: Two promises are OK I think, they are different targets. or did you mean something else?

Comment (Lex-2008) [reply cfe_internal/update/update_policy.cf:244]: I mean that if they don't get executed in proper order, there might be an agent run when a policy is updated, but the modules are not there.

File: cfe_internal/update/update_policy.cf (hunk 3 above)
Comment (Lex-2008) [reply cfe_internal/update/update_policy.cf:246]: from https://docs.cfengine.com/docs/master/reference-promise-types-files.html#mode:

> Notes: This is ignored on Windows, as the permission model uses ACLs.

Also note that Windows doesn't have "executable" permission bit, but checks file extensions instead. So if you can read an `*.exe` file - you can as well execute it.

**Comment by craigcomstock** [comment]
This commit breaks agents running updates. PR in the works to fix.

**Comment by craigcomstock** [comment]
On second thought, my quick-fix PR wasn't fully baked, so will revert this.

Failed cases in CI from this PR are I believe:

- [ ] bootstrap-tests
  - [ ] new_tests/bootstrap_tests/.correct_permissions_after_bootstrap.test
  - [ ] new_tests/bootstrap_tests/.restart.test
- [ ] sequential-tests
  - [ ] new_tests/sequential_tests/01-basic.01-policy_distribution.test
  - [ ] new_tests/sequential_tests/02-upgrade-minor.01-upgrade.test

**Pull request merged**